### PR TITLE
Ledger login: enable custom HD path

### PR DESCRIPTION
**Pull request by Patrick1904**
Changes:
1. When logging in with Ledger, it shows an 'advanced options' dropdown that allows the user to modify the ledger HD path

### CODE REVIEW

**Comment by Patrick1904** [comment]
Some follow-up improvements suggested by @jimmy3dita:
1. have the collapsible close automatically when I click "select HD path"
2. have the advanced option to select the HD path also when I create a new account
3. don't create a new key when I login with Ledger, to keep the process stealthy and the account "clean"

**Review by jimmy3dita — APPROVED** [review]
I was able to generate an `unknown error` if I leave a Ledger X plugged and it goes on standby. However, waking it up, typing the PIN, and clicking "retry" completes the procedure.
```
Uncaught (in promise) 
Qe
message: "Ledger device: UNKNOWN_ERROR (0x6804)"
name: "TransportStatusError"
stack: "Error↵    at new Qe (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12181)↵    at g.send (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:1091)"
statusCode: 26628
statusText: "UNKNOWN_ERROR"
__proto__: Error at Object.parcelRequire.l0DP (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12270) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.zld1.events (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:571) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.yJzG.u2f-api (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:20:139) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.StEj.near-ledger-js (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:22:137) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at f (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:233) at p (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:544) at https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1080:250
stack: "Error↵    at Object.parcelRequire.l0DP (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12270)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.zld1.events (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:571)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.yJzG.u2f-api (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:20:139)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.StEj.near-ledger-js (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:22:137)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at f (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:233)↵    at p (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:544)↵    at https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1080:250"
__proto__:
constructor: ƒ Error()
message: ""
name: "Error"
toString: ƒ toString()
__proto__: Object
```

**Comment by Patrick1904** [comment]
> I was able to generate an `unknown error` if I leave a Ledger X plugged and it goes on standby. However, waking it up, typing the PIN, and clicking "retry" completes the procedure.
> 
> ```
> Uncaught (in promise) 
> Qe
> message: "Ledger device: UNKNOWN_ERROR (0x6804)"
> name: "TransportStatusError"
> stack: "Error↵    at new Qe (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12181)↵    at g.send (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:1091)"
> statusCode: 26628
> statusText: "UNKNOWN_ERROR"
> __proto__: Error at Object.parcelRequire.l0DP (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12270) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.zld1.events (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:571) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.yJzG.u2f-api (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:20:139) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544) at Object.parcelRequire.StEj.near-ledger-js (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:22:137) at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468) at f (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:233) at p (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:544) at https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1080:250
> stack: "Error↵    at Object.parcelRequire.l0DP (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:12:12270)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.zld1.events (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:15:571)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.yJzG.u2f-api (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:20:139)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at p (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:544)↵    at Object.parcelRequire.StEj.near-ledger-js (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:22:137)↵    at f (https://near-wallet-pr-1559.onrender.com/ledger.9d275b66.js:1:468)↵    at f (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:233)↵    at p (https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1:544)↵    at https://near-wallet-pr-1559.onrender.com/src.958761a4.js:1080:250"
> __proto__:
> constructor: ƒ Error()
> message: ""
> name: "Error"
> toString: ƒ toString()
> __proto__: Object
> ```

Yes it can happen if it's left idle. Refreshing the page should solve it. Btw I fixed so that we close the dropdown on 'confirm' now - it just makes sense.

**Comment by jimmy3dita** [comment]
Triggered another error: if I use and implicit account, the button "send" will return a blank page. 
From the console:
```
TypeError: Cannot read property 'addEventListener' of null
    at Accordion.js:26
    at ko (react-dom.production.min.js:211)
    at ju (react-dom.production.min.js:257)
    at exports.unstable_runWithPriority (scheduler.production.min.js:19)
    at ri (react-dom.production.min.js:122)
    at Hu (react-dom.production.min.js:257)
    at Cu (react-dom.production.min.js:239)
    at react-dom.production.min.js:123
    at exports.unstable_runWithPriority (scheduler.production.min.js:19)
    at ri (react-dom.production.min.js:122)
    at oi (react-dom.production.min.js:123)
    at ai (react-dom.production.min.js:122)
    at F (react-dom.production.min.js:287)
    at tn (react-dom.production.min.js:68)
    at HTMLDocument.s (helpers.ts:87)
(anonymous) @ instrument.ts:129
helpers.ts:111 Uncaught TypeError: Cannot read property 'addEventListener' of null
    at Accordion.js:26
    at ko (react-dom.production.min.js:211)
    at ju (react-dom.production.min.js:257)
    at exports.unstable_runWithPriority (scheduler.production.min.js:19)
    at ri (react-dom.production.min.js:122)
    at Hu (react-dom.production.min.js:257)
    at Cu (react-dom.production.min.js:239)
    at react-dom.production.min.js:123
    at exports.unstable_runWithPriority (scheduler.production.min.js:19)
    at ri (react-dom.production.min.js:122)
    at oi (react-dom.production.min.js:123)
    at ai (react-dom.production.min.js:122)
    at F (react-dom.production.min.js:287)
    at tn (react-dom.production.min.js:68)
    at HTMLDocument.s (helpers.ts:87)
```

**Comment by Patrick1904** [comment]
@vgrichina Fossa keeps failing for some reason unrelated to this PR. @chefsale fixed it earlier when it was failing by restarting the test. I tried restarting a couple of times this time but doesn't seem to help.

Screen Shot 2021-03-22 at 4 53 32 PM

**Comment by Patrick1904** [comment]
> Triggered another error: if I use an implicit account [0fe102696024c6c32e22ad2480a9dcb481094df2d86e897d4fef20d7bd010348](https://explorer.testnet.near.org/accounts/0fe102696024c6c32e22ad2480a9dcb481094df2d86e897d4fef20d7bd010348), clicking the button "send" will return a blank page.

Thanks for catching @jimmy3dita! Will fix.

**Comment by jimmy3dita** [comment]
On Mainnet we have these two errors:
```
Error in background action: Error: [-32000] Server error: Account ID '' is not valid
    at json-rpc-provider.js:179
    at async Object.e [as default] (exponential-backoff.js:7)
    at async f.sendJsonRpc (json-rpc-provider.js:154)
    at async f.query (json-rpc-provider.js:77)
    at async g.state (account.js:55)
(anonymous) @ instrument.ts:129
deploy-preview-1559-…ging.netlify.app/:1

Uncaught (in promise) Error: [-32000] Server error: Account ID '' is not valid
    at json-rpc-provider.js:179
    at async Object.e [as default] (exponential-backoff.js:7)
    at async f.sendJsonRpc (json-rpc-provider.js:154)
    at async f.query (json-rpc-provider.js:77)
    at async g.state (account.js:55)
    ```

**Comment by Patrick1904** [comment]
> On Mainnet we have these two errors:
> 
> ```
> Error in background action: Error: [-32000] Server error: Account ID '' is not valid

@jimmy3dita this is known. It's a check for logged in account (not related to this PR)

**Comment by Patrick1904** [comment]
> On Mainnet we have these two errors using the path `44'/397'/0'/0'/3'` (implicit account `b88a2f4ca8c28e48b8f3e60f5a96b4ca0c03ad5312e036c189ed7602402fb74d`):
> 
> ```
> Error in background action: Error: [-32000] Server error: Account ID '' is not valid
>     at json-rpc-provider.js:179
>     at async Object.e [as default] (exponential-backoff.js:7)
>     at async f.sendJsonRpc (json-rpc-provider.js:154)
>     at async f.query (json-rpc-provider.js:77)
>     at async g.state (account.js:55)
> (anonymous) @ instrument.ts:129
> deploy-preview-1559-…ging.netlify.app/:1

@jimmy3dita it seems RPC is not able to find the account ID. How did you trigger this?

**Comment by Patrick1904** [comment]
@chefsale I tried restarting the Fossa test a few times again like you did to solve the test failure, but it doesn't seem to fix it. Do you have any idea what might be the actual issue? I don't see how it's related to any change I made

Screen Shot 2021-03-22 at 5 11 24 PM

**Comment by jimmy3dita** [comment]
> @jimmy3dita it seems RPC is not able to find the account ID. How did you trigger this?

1. create an implicit account via https://near-ledger-keygen.onrender.com/
2. fund the address sending some tokens (e.g. [tx EU1qUZrFRtryLAUN5kAqxJBXuEnHQr13ErYPHT9nmYGw](https://explorer.near.org/transactions/EU1qUZrFRtryLAUN5kAqxJBXuEnHQr13ErYPHT9nmYGw))
3. recover account via ledger path using the provided staging link
4. click "profile" or "send". 

This is happening only if I open a new "incognito window" on Chrome, not on the cached version. Seems to me it's happening if there are no named accounts in the top-right account selection menu

**Review by jimmy3dita — APPROVED** [review]

**Comment by Patrick1904** [comment]
@vgrichina this is ready. Please take a look when you have some time.

File: src/components/profile/hardware_devices/HardwareDevices.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ const HardwareDevices = ({ recoveryMethods }) => {
     const publicKeys = keys.map(key => key.public_key)
     const hasOtherMethods = publicKeys.some(key => recoveryKeys.includes(key))
     const hasLedger = userRecoveryMethods.filter(method => method.kind === 'ledger').map(key => key.publicKey).some(key => publicKeys.includes(key))
-    const ledgerIsConnected = account.ledgerKey !== null && hasLedger;
```
Comment (vgrichina) [inline src/components/profile/hardware_devices/HardwareDevices.js:73]: @Patrick1904 why is this change needed, is info about `ledger` not added correctly to recovery methods?

File: src/components/common/Accordion.js [hunk 2]
```diff
@@ -5,25 +5,34 @@ import classNames from '../../utils/classNames'
...
 `
 
-const Accordion = ({ className, trigger, children }) => {
+const Accordion = ({ className, trigger, children, transition = '250' }) => {
     const [open, setOpen] = useState(false)
     const [contentHeight, setContentHeight] = useState('')
 
     useEffect(() => {
         const el = document.getElementById(trigger)
+        const secondEl = document.getElementById(`${trigger}-2`)
```
Comment (vgrichina) [inline src/components/common/Accordion.js:23]: @Patrick1904 nit: maybe it makes to do something for all events with some class vs having code explicitly only for 1 or 2 elements?

**Review by vgrichina — APPROVED** [review]
@Patrick1904 Looks good overall, but I have few questions

**Comment by vgrichina** [comment]
@jimmy3dita 

> P.s.: it's painful to see this proliferation of authorized keys every time I use the "recover" button from an incognito Chrome window 😅

We can put this on a roadmap:
https://github.com/near/near-wallet/issues/772

I also think we need to have key management re-design on roadmap. Like we definitely need an easy UI to do mass key cleanup for all your unused apps and devices in one transaction. So that you don't have to cleanup every key manually.

File: src/components/profile/hardware_devices/HardwareDevices.js (hunk 1 above)
Comment (Patrick1904) [reply src/components/profile/hardware_devices/HardwareDevices.js:73]: @vgrichina since the account can be secured with Ledger outside of the NEAR Wallet, it's not a guarantee that contract helper as the record. So if the user logs in with Ledger in the NEAR Wallet, we need to show it as 'active' despite no record in contract helper. 

In general it seems to make more sense too to show Ledger as enabled/connected if there's a Ledger key present, since that's the ultimate truth.

**Comment by Patrick1904** [comment]
@vgrichina the fossa test failing sporadicly again but shouldn't be related to the last commit.

File: src/utils/wallet.js [hunk 3]
```diff
@@ -100,9 +100,10 @@ class Wallet {
             async signMessage(message, accountId, networkId) {
```
Comment (vgrichina) [inline src/utils/wallet.js:100]: @Patrick1904 need to use HD path in getPublicKey above as well

**Review by vgrichina — APPROVED** [review]
@Patrick1904 Looks good overall except getPublicKey

File: src/utils/wallet.js (hunk 3 above)
Comment (Patrick1904) [reply src/utils/wallet.js:101]: @vgrichina please check again after our discussion (only device functions take path, like `client.getPublicKey` or `client.sign`)
